### ISSUE TITLE
Search engine blocks

### ISSUE THREAD

**Issue by 737simpilot** [CIDRAM/CIDRAM#207]
I get these search engine blocks and I don't understand why. I'll go over each one.

```
ID: 1617264971-385388-6230495907
Date/Time: Thu, 01 Apr 2021 02:16:11 -0600
IP Address: 20.191.45.212
Referrer: https://my_web_site.com/
Signatures Count: 1
Signatures Reference: 20.188.0.0/14
Why Blocked: Cloud service ("msaz_cidram.inc-IPv4", L492:F11)!
User Agent: Mozilla/5.0 (compatible; DuckDuckGo-Favicons-Bot/1.0; +http://duckduckgo.com)
Reconstructed URI: https://my_web_site.com/
reCAPTCHA State: Disabled.
```

That IP belongs to DuckDuckGo according to the IP listed in the verification.yaml file. I ran the IP and UA in the IP Test page and it says, "Yes – Cloud service ("msaz_cidram.inc-IPv4", L492:F11)."  Perhaps the unsuccessful querying of a PTR record is to blame? I can't see why since that IP is in verification.yaml

```
ID: 1617273564-783942-4712631924
Date/Time: Thu, 01 Apr 2021 04:39:24 -0600
IP Address: 40.77.139.34
Query: v=2180135197
Referrer: https://my_web_site.com/
Signatures Count: 1
Signatures Reference: 40.77.138.0/23
Why Blocked: Cloud service ("msaz_cidram.inc-IPv4", L789:F11)!
User Agent: Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/534+ (KHTML, like Gecko) BingPreview/1.0b
Reconstructed URI: https://my_web_site.com/app.php/extrastyle/showtopicstarter__css?v=2180135197
reCAPTCHA State: Disabled.
```

This one _could_ be fake, and there is no PTR record. I assume it's Microsoft's crap with a search bot on an IP with no PTR record.

```
ID: 1617312298-153061-8332631018
Date/Time: Thu, 01 Apr 2021 15:24:58 -0600
IP Address: 207.46.13.95
Query: t=52
Signatures Count: 1
Signatures Reference: 207.46.13.0/24
Why Blocked: Cloud service ("msaz_cidram.inc-IPv4", L2098:F11)!
User Agent: Mozilla/5.0 (compatible; bingbot/2.0; +http://www.bing.com/bingbot.htm)
Reconstructed URI: https://my_web_site/viewtopic.php?t=52
reCAPTCHA State: Disabled.
```

This one is really interesting. It has a PTR record, and the IP Test page for the IP and UA combination says it isn't blocked, but as we can see here it was in fact blocked. 

Now I don't know if this is a Macmathan issue or a CIDRAM issue or the combination of both. Looking in the Cache Data I see all three of these IPs in the DNS Reverses list. 

Edit-

I see in my access logs that other than these blocks, Bing bot is getting through. I rarely get DuckDuckGo hits. I think it's because I was blocking them for so long at the CloudFlare layer. I finally whitelisted their IPs and now CIDRAM is blocking them.

**Comment by Maikuolan** [comment]
Some of these, I'm not immediately sure what's going on. I'll need to investigate each one individually, and I may need to reply here in multiple parts, as I get some results for each one.

That last one there: Definitely odd! At my end, I can confirm slightly different, but equally weird results. For me, the IP test page is showing that IP as not being blocked, and faking an actual request from that IP address, it successfully gets through, confirming what's reported by the IP test page. However, I can quite clearly see a signature in `msaz_cidram.inc`, which *should* be blocking it:

`207.46.13.0/24 Deny Cloud`

So, whether a false positive or otherwise, I'm not sure why it isn't being blocked at my end currently. I can't immediately see any anything obviously wrong with the signature file or any mistakes in it (in terms of syntax, formatting, etc).

Regarding the very first one: That looks like a legitimate request from DuckDuckGo, so the block was definitely a false positive and shouldn't have occurred. The PTR record isn't to blame there though. It seems that, though a legitimate request, said IP address also happens to be a part of the Azure cloud, thus justifiably blocked by the Azure signatures file. The current settings for DuckDuckGo in the search engine and social media verification file prevent any legitimate DuckDuckGo IPs from ever being *banned* by CIDRAM, but don't necessarily prevent those IPs from being *blocked*, if there are any signatures anywhere which would normally cause a block event to occur. I guess, in light of the discover that one of their IPs belongs to Azure, I may need to revisit those settings, to account for that and to prevent it from being blocked, too.

**Comment by 737simpilot** [comment]
In the IP Test page, testing the IP `207.46.13.95` with a UA of: `Mozilla/5.0 (Windows NT 10.0; WOW64; rv:87.0) Gecko/20100101 Firefox/87.0` says: `Yes – Cloud service ("msaz_cidram.inc-IPv4", L2098:F11)`

Testing the IP again with a UA of: `Mozilla/5.0 (compatible; bingbot/2.0; +http://www.bing.com/bingbot.htm)` says: `No.`

But if I check "Also test against auxiliary rules." I get: `Yes – Cloud service ("msaz_cidram.inc-IPv4", L2098:F11), Bing MSAZ Bypass, Ref Block, Security ++Mark for use with reCAPTCHA`

The Bing MSAZ Bypass is mentioned on line 30 of the xcldbyp_cidram.inc signuture file. The Ref Block, Security refer to my two auxiliary rules. The Ref block just says if a referrer isn't from an HTTPS URL, then block. And the Security auxiliary rule blocks all POST requests except those I specify which are pertaining to the website, and all other URLs for a POST request would be blocked.

I just verified my Security auxiliary rule, and viewtopic.php?* is whitlelisted for a POST request. So if Bing bot or what ever on that server made a POST request, it should go through unless other Sigs trigger. 

Question: I want to test with and without the auxiliary rules. Can I pull out the auxiliary rules straight from the yaml file and paste them back when done? Or is there some other mechanism I need to contend with?

**Comment by Maikuolan** [comment]
> Question: I want to test with and without the auxiliary rules. Can I pull out the auxiliary rules straight from the yaml file and paste them back when done?

Yep.

> Or is there some other mechanism I need to contend with?

Nope.

All information pertaining to the auxiliary rules is contained by the auxiliary rules file, so editing just that file alone should be sufficient for manually editing/adding/deleting any auxiliary rules. :-)

**Comment by 737simpilot** [comment]
Thanks. I thought as much. I'll do some testing here without auxiliary rules just for giggles and see if there's a change in the IP Tracking.

**Comment by Maikuolan** [comment]
No further progress on this yet due to lack of time. I'll get there soon though. :-)

**Comment by 737simpilot** [comment]
No problem. Right now I just whitelistd the DuckDuckGo IPs in the auxiliary rules and have don't log for the Azure hits. For the Azure hits it doesn't happen all that often. I think the DuckDuckGo hits are more often though.  It's looking like the common denominator is that the Azure IPs don't go back to a PTR. So this could be either Microsoft's mess up or it's some hacker bot what ever which I have my doubts on. The IP I have there of 40.77.139.34 in my opening post was blocked, and there is no PTR that I can see running the Windows command of ping -a 40.77.139.34. So naturally that would be blocked I guess for lack of rDNS, but checking the IP 207.46.13.95 listed in my opening post DOES have a PTR according the the ping -a switch command, yet it was blocked.

**Comment by Maikuolan** [comment]
`20.191.45.212` and `207.46.13.95` shouldn't be blocked anymore now. `40.77.139.34` is still being blocked though, since it's failing verification to begin with. Could be right about it just being another case of Microsoft not resolving/reversing the IP properly due to bad/missing PTR records. Not sure whether we could, or even should try to solve that though.

**Comment by Maikuolan** [comment]
Thoughts? @737simpilot

**Comment by 737simpilot** [comment]
So, over time now I've seen lots of Azure IPs that do look and act like Bing bot, but their PTR never verifies that. So I don't think Microsoft is not doing what they should be doing and creating the correct PTR for their bot.  It is within the realm of possibility the IP hits are bogus, but we can't be sure. Although, I've seen Bing bot probe JS resources. Not sure why they'd wanna do that. Maybe Google and the others do the same thing. I'd have to check my access logs. I should know this. It's on me to see what resources are being looked at and by who. For the most part I do look at that, but if I know the IP is Google or Bing, etc I skip past that log entry. Probably not a good idea on my part even though it _may_ be legit.

40.77.139.34 doesn't have a PTR, and I checked at AbuseIPDB and one lone person has two pages worth of reports related to some port scan. Seen as how that's just one and only one person and that I've seen people report Google bot of all things, I'd call this within the realm of conjecture to be listed at all at AbuseIPDB. My thoughts on keeping it blocked? Probably a good idea. What's one possible legit IP out of how many? Bing et al can still probe the website anyhow, I don't think it'll dock the SEO points so to speak. And as I said, it seems like Microsoft is _soft_ on not adding proper PTR records. And to make matters worse, they intermingle their bot with Azure so it's hard to differentiate between the crap. But to be fair, Google does the same thing. The common denominator is that I don't think Google or Microsoft uses just one or more CIDRs dedicated to bot traffic so then we can just flat out whitelist or greylist those CIDRs. We're left with having to verify PTRs. And Google themselves says that's what we should do. Why not just put the damn bot traffic in a CIDR or a list of dedicated IPs? Ugh... At least DuckDuckGo does. This is just just like Google's HSTS pre-load list. They use a hodgepodge of IPs from Google Cloud and a crap UA. So for me it's next to impossible to let it hit my site so I'm still on the list for HSTS for Chrome, Firefox et al that use the list and at the same time keep the crap at bay. If I could IP/UA match the traffic for whitelisting it wouldn't be an issue. This pre-load list is a load of crap anyway. It should be done via the blockchain, and there should be a dedicated CIDR or IPs for it or command your search engine bot to determine HSTS validation. Moving on... LOL

I don't see a PTR for 20.191.45.212. And yet again, the incredible low IQ morons who probably shouldn't be running a server let alone a website have reported this IP at AbsueIPDB and the most laughable is the reported "fake" DuckDuckGo bot. LMAO!  If anyone doesn't understand the validity here in what I said, check CIDRAM's verification.yaml file for the IP and here: https://duckduckgo.com/duckduckbot

207.46.13.95 does have a PTR matching that of Bingbot. 

The two methods I use to verify the PTR are this site: https://ipx.ac/

And entering the command `ping -a x.x.x.x` The x's being the IP.  This command works in Windows command prompt. If you're in command prompt and simply type `ping` and hit enter you'll see all the switches. the -a switch is to resolve host names and it matches what the ipx.ac website reports. 

So I was wondering. How are you confirming the PTR? If you're just going by a reverse DNS then I _think_ that would just give you a name server name, not the actual pointer record we are after. Or am I wrong? 

Another question:  Do you know if a person can run his/her own VPS or bare metal server and create their own PTR? I've been thinking about this and if someone were to buy some server resources at Azure or Google which now means your in that ASN and then craft a PTR mimicking a search engine, now they have a path in to do damage. Or is the PTR setup outside of the server resource by the host themselves so it's not changeable via the VPS/server customer? I really don't know. I don't run a VPS right now because I just don't have the traffic to merit the cost. I want to though for the added security on my part with IPtables, Fail2Ban, Snort and all the rest. It's analogous to the CID (Caller ID) in a DiD (Direct Inward Dialing) number for a VoIP service.  Anyone can just buy some VoIP service and change the CID to what ever they like. So I could mimic the Visa financial services corporation, call you and say there's been an error on your credit card yada yada. I ask for you to verify your credit card number, expiration date and CVV and now I just ripped you off.  So yeah, I was thinking the PTR could in theory be changed by the VPS customer like a CID so our so-called verification is potentially out the window.

**Comment by 737simpilot** [comment]
Interesting. I said the following in my opening post for this issue:


> This one could be fake, and there is no PTR record. I assume it's Microsoft's crap with a search bot on an IP with no PTR record.

>ID: 1617312298-153061-8332631018
>Date/Time: Thu, 01 Apr 2021 15:24:58 -0600
>IP Address: 207.46.13.95
>Query: t=52
>Signatures Count: 1
>Signatures Reference: 207.46.13.0/24
>Why Blocked: Cloud service ("msaz_cidram.inc-IPv4", L2098:F11)!
>User Agent: Mozilla/5.0 (compatible; bingbot/2.0; +http://www.bing.com/bingbot.htm)
>Reconstructed URI: https://my_web_site/viewtopic.php?t=52
>reCAPTCHA State: Disabled.

Now I see this IP does have a PTR that matches Bingbot. :roll_eyes: They must have finally updated the code in the _Matrix_...

**Comment by Maikuolan** [comment]
> Do you know if a person can run his/her own VPS or bare metal server and create their own PTR? I've been thinking about ...

Good question. Should be possible in *theory*, I imagine, though it would likely depend on exactly how much control the person has over that server, as well as depending on quite a few other factors playing out in that person's favour in order for it to be exploitable, or to affect others across the internet more broadly (e.g., peering to the network in question, routes, etc). TBH though, I don't have a huge amount of experience dealing with PTR records. I might forward this on to a few others I know, to see whether they'd be interested in chiming in.

**Comment by 737simpilot** [comment]
So, I read [this](https://www.computerworld.com/article/2833006/how-to-setup-reverse-dns-and-ptr-records.html) old post, I see that it's possible? I mean, that relates to email servers, but none the less could be done in a Google cloud VPS where that CIDR is shared among the Googlebot PTRs we are verifying. That's my theory anyway.

Check this out: https://www.google.com/search?client=firefox-b-1-d&q=how+to+change+ptr

Look at the "people also ask" drop downs and their sources.

**Comment by Maikuolan** [comment]
Just a quick note: I reread over this issue just now, and double-checked all the affected IPs just now to see whether there'd been any changes (e.g., lookups suddenly working correctly now or something like that), but everything still looks the same as it did back in May, when we were last discussing this issue. So, issue still valid, and still work needed to be done.

**Comment by 737simpilot** [comment]
![img](https://i.postimg.cc/3wzyfSwD/ezgif-com-gif-maker-6.gif)

**Comment by Maikuolan** [comment]
For better or worse, for whatever the reason.. Since my previous reply here, I hadn't done anything about this issue at all or done anything specifically with the intention of progressing it at all, but double-checked all the affected IPs again just now to know whether anything might've changed either at their end or inadvertently at our end, and.. it seems that none of the affected IPs are blocked anymore.

That being the case.. I might be able to mark this issue as resolved now, I guess..? @737simpilot Mind confirming at your end when you get a moment? Cheers.

**Comment by 737simpilot** [comment]
Yeah, that's fine. I've been busy and whatnot and actually forgot I opened this issue and had to reread over everything again. I eventually got tired of Microsoft's very aggressive bot (there's at least 30% more traffic from MS than Google) and their ASN also being a real PITA with Azure shenanigan IPs, that I just out right completely blocked all of their ASNs at the Cloudflare level. 

And with DuckDuckGo being partners with Microsoft, and being a PITA to whitelist their changing IPs all the time in Cloudflare, and how they just load my index and nothing more, I'm about to block them too. I don't think it'll matter much either. I just allow Google. So DuckDuckGo et al use Google anyway... Even via Searx. (Thinking of spinning up my own private Searx search engine one day actually).

So yeah, for me personally I'll call this solved. I don't think it applies to anyone else. Especially since no one else chimed in during a year's time frame.
